src/notification.c

```diff
@@ -347,12 +347,19 @@ check_notification (GVariant *notification,
 }
 
 static GVariant *
-maybe_remove_icon (GVariant *notification)
+sanitize_notification (XdpAppInfo *app_info,
+                       GVariant   *notification)
 {
+  const char *desktop_id;
   GVariantBuilder n;
   int i;
 
   g_variant_builder_init (&n, G_VARIANT_TYPE_VARDICT);
+
+  desktop_id = xdp_app_info_get_desktop_id (app_info);
+  if (desktop_id && g_strcmp0 (xdp_app_info_get_id (app_info), desktop_id) != 0)
+    g_variant_builder_add (&n, "{sv}", "desktop-id", g_variant_new_string (desktop_id));
+
   for (i = 0; i < g_variant_n_children (notification); i++)
```

Comment:
The `desktop-id` field is being set on the call from the portal to the backend and the value comes from snapd so it is correct. There's no `desktop-id` field received from the call from the app to the portal, so this can't be set to an invalid value.

Reply:
I thought the portal was copying the contents of the `a{sv}` from the caller as-is.

I now see that `check_notification()` only allows certain keys, and rejects any call that has others (and in particular it would reject a notification that has a `desktop-entry` or `desktop-id`. If this is a guarantee, then I'm still a little uncomfortable about mixing trusted and untrusted fields in the same dict, but we don't strictly need new API.

Reply:
@smcv yeah, I had more or less the same concern.

So, while using `impl-` prefix could help avoid such stuff to be "leaked" to the app, it still depends on implementer.

But the alternative was to add another method call `AddNotification2` with platform/private data, and it seemed also too much to me for just this.